File: Assets/scripts/Packets.cs
using System;
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using Newtonsoft.Json;

public enum PcktType
{
    Error = 0,
    ClientHello,
    ServerHello,
    PlayerInfo,
    PlayerEnter,
    PlayerLeave,
    PlayerStats,
    Attack,
    Shoot
}

public class Packet
{
    public PcktType type;
    public int id;
    public string ToJson(int netId)
    {
        id = netId;
        return JsonConvert.SerializeObject(this);
    }
}

public class ClientHello : Packet
{
    public ClientHello()
    {
        type = PcktType.ClientHello;
    }
}

public class ServerHello : Packet
{
    public ServerHello()
    {
        type = PcktType.ServerHello;
    }
}

public class PlayerInfo : Packet
{
    public PlayerInfo()
    {
        type = PcktType.PlayerInfo;
    }
    //public Position position;
    public float x, y, rotation;
    public void loadFromTransform(Transform t)
    {
        x = t.position.x;
        y = t.position.y;
        rotation = t.rotation.z;
    }
}


public class PlayerEnter : Packet
{
    public PlayerEnter()
    {
        type = PcktType.PlayerEnter;
    }
    public int gid;
}

public class PlayerLeave : Packet
{
    public PlayerLeave()
    {
        type = PcktType.PlayerLeave;
    }

}

public class Shoot : Packet
{
    public float start_x, start_y, end_x, end_y;
    public Shoot()
    {
        type = PcktType.Shoot;
    }
}

public class PlayerStats : Packet
{
    public PlayerStats()
    {
        type = PcktType.PlayerStats;
    }
    public float health;
}

File: server/gamelib/packets.py

from enum import IntEnum, auto
from json import load as json_load
from pydantic import BaseModel, Field,RootModel
from typing import Literal, Union
from .libmath import Vector2

class MsgType(IntEnum):
    Error = 0
    ClientHello = auto()
    ServerHello = auto()
    PlayerInfo = auto()
    PlayerEnter = auto()
    PlayerLeave = auto()
    PlayerStats = auto()
    Attack = auto()
    Shoot = auto()
    PlayerWeapon = auto()


class MsgError(BaseModel):
    type: Literal[MsgType.Error] = MsgType.Error
    msg: str


class MsgClientHello(BaseModel):
    type: Literal[MsgType.ClientHello] = MsgType.ClientHello
    


class MsgServerHello(BaseModel):
    type: Literal[MsgType.ServerHello] = MsgType.ServerHello
    name: str
    id: int


class MsgPlayerInfo(BaseModel):
    type: Literal[MsgType.PlayerInfo] = MsgType.PlayerInfo
    id: int
    x: float
    y: float
    rotation: float

class MsgPlayerEnter(BaseModel):
    type: Literal[MsgType.PlayerEnter] = MsgType.PlayerEnter
    id: int
    gid: int

class MsgPlayerLeave(BaseModel):
    type: Literal[MsgType.PlayerLeave] = MsgType.PlayerLeave
    id: int

class MsgPlayerStats(BaseModel):
    type: Literal[MsgType.PlayerStats] = MsgType.PlayerStats
    id: int
    health: float


class MsgAttack(BaseModel):
    type: Literal[MsgType.Attack] = MsgType.Attack
    id: int
    target: int

class MsgShoot(BaseModel):
    type: Literal[MsgType.Shoot] = MsgType.Shoot
    id: int
    start_x: float
    start_y: float
    end_x: float
    end_y: float
    
class MsgPlayerWeapon(BaseModel):
    type: Literal[MsgType.PlayerWeapon] = MsgType.PlayerWeapon
    id: int
    weapon_id: int
    

class Msg(RootModel):
    root: Union[MsgError, MsgPlayerInfo, MsgPlayerStats, MsgAttack, MsgShoot, MsgPlayerWeapon] = Field(..., discriminator="type")
